File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/DoubleEntityTransform.java
package com.loudsight.meta.serialization.transform;

import com.loudsight.meta.serialization.EntityTransform;
import com.loudsight.meta.serialization.EntityType;

import java.util.Iterator;
import java.util.List;

public class DoubleEntityTransform extends EntityTransform<Double> {

    private static class DoubleEntityTransformHolder {
        private static final DoubleEntityTransform INSTANCE = new DoubleEntityTransform();
    }
    // global access point
    public static DoubleEntityTransform getInstance() {
        return DoubleEntityTransform.DoubleEntityTransformHolder.INSTANCE;
    }
    private DoubleEntityTransform() {
        super(EntityType.DOUBLE, Double.class);
    }


    @Override public void serializeEntity( Double entity, List<Byte> bytes) {
        bytes.add(EntityType.DOUBLE.getCode());
        writeLong(Double.doubleToRawLongBits(entity), bytes);
    }

    @Override public Double  deserializeEntity(Iterator<Byte> bytes) {
        var doubleBits = readLong(bytes);

        return Double.longBitsToDouble(doubleBits);
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/transform/DoubleEntityTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.serialization.EntityTransform
import com.loudsight.meta.serialization.EntityType

object DoubleEntityTransform : EntityTransform<Double>(EntityType.DOUBLE, Double::class) {

    override fun serializeEntity(entity: Double, bytes: MutableList<Byte>) {
        bytes.add(EntityType.DOUBLE.code)
        writeLong(entity.toBits(), bytes)
    }

    override fun deserializeEntity(bytes: Iterator<Byte>): Double {
        val doubleBits = readLong(bytes)

        return Double.fromBits(doubleBits)
    }


    override fun canTransform(entity: Any): Boolean {
        return entity is Double
    }

}